File: source_code/app/models/schemas.py
from pydantic import BaseModel, EmailStr, Field
from typing import Optional, List
from datetime import datetime


# ── Auth Schemas ──────────────────────────────────────────────
class UserRegister(BaseModel):
    name: str = Field(..., min_length=2, max_length=100)
    email: EmailStr
    password: str = Field(..., min_length=6)


class UserLogin(BaseModel):
    email: EmailStr
    password: str


class UserOut(BaseModel):
    id: str
    name: str
    email: str
    created_at: datetime


class Token(BaseModel):
    access_token: str
    token_type: str = "bearer"


# ── Product Schemas ───────────────────────────────────────────
class ProductCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    description: str
    price: float = Field(..., gt=0)
    category: str
    image_url: Optional[str] = None
    stock: int = Field(default=0, ge=0)
    tags: Optional[List[str]] = []


class ProductOut(BaseModel):
    id: str
    name: str
    description: str
    price: float
    category: str
    image_url: Optional[str]
    stock: int
    tags: List[str]
    rating: Optional[float] = 0.0
    created_at: datetime


# ── Cart Schemas ──────────────────────────────────────────────
class CartItem(BaseModel):
    product_id: str
    quantity: int = Field(..., gt=0)


class CartItemOut(BaseModel):
    product_id: str
    name: str
    price: float
    quantity: int
    subtotal: float


class CartOut(BaseModel):
    user_id: str
    items: List[CartItemOut]
    total: float


# ── Order Schemas ─────────────────────────────────────────────
class OrderOut(BaseModel):
    id: str
    user_id: str
    items: List[CartItemOut]
    total: float
    status: str
    created_at: datetime


# ── Review Schemas ────────────────────────────────────────────
class ReviewCreate(BaseModel):
    product_id: str
    rating: int = Field(..., ge=1, le=5)
    comment: Optional[str] = None


class ReviewOut(BaseModel):
    id: str
    user_id: str
    product_id: str
    rating: int
    comment: Optional[str]
    created_at: datetime

File: source_code/app/models/schemas.js
const mongoose = require('mongoose');

// ── User ──────────────────────────────────────────────────────
const userSchema = new mongoose.Schema({
  name:       { type: String, required: true, minlength: 2, maxlength: 100 },
  email:      { type: String, required: true, unique: true, lowercase: true },
  password:   { type: String, required: true },
  isAdmin:    { type: Boolean, default: false },
}, { timestamps: true });

// ── Product ───────────────────────────────────────────────────
const productSchema = new mongoose.Schema({
  name:        { type: String, required: true, maxlength: 200 },
  description: { type: String, required: true },
  price:       { type: Number, required: true, min: 0 },
  category:    { type: String, required: true },
  image_url:   { type: String, default: null },
  stock:       { type: Number, default: 0, min: 0 },
  tags:        { type: [String], default: [] },
  rating:      { type: Number, default: 0 },
}, { timestamps: true });

// ── Cart ──────────────────────────────────────────────────────
const cartItemSchema = new mongoose.Schema({
  product_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Product', required: true },
  quantity:   { type: Number, required: true, min: 1 },
}, { _id: false });

const cartSchema = new mongoose.Schema({
  user_id: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, unique: true },
  items:   { type: [cartItemSchema], default: [] },
}, { timestamps: true });

// ── Order ─────────────────────────────────────────────────────
const orderItemSchema = new mongoose.Schema({
  product_id: { type: String, required: true },
  name:       { type: String, required: true },
  price:      { type: Number, required: true },
  quantity:   { type: Number, required: true },
  subtotal:   { type: Number, required: true },
}, { _id: false });

const orderSchema = new mongoose.Schema({
  user_id: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  items:   { type: [orderItemSchema], default: [] },
  total:   { type: Number, required: true },
  status:  { type: String, default: 'pending' },
}, { timestamps: true });

// ── Review ────────────────────────────────────────────────────
const reviewSchema = new mongoose.Schema({
  user_id:    { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  product_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Product', required: true },
  rating:     { type: Number, required: true, min: 1, max: 5 },
  comment:    { type: String, default: null },
}, { timestamps: true });

module.exports = {
  User:    mongoose.model('User',    userSchema),
  Product: mongoose.model('Product', productSchema),
  Cart:    mongoose.model('Cart',    cartSchema),
  Order:   mongoose.model('Order',   orderSchema),
  Review:  mongoose.model('Review',  reviewSchema),
};
